### ISSUE TITLE
RFC Process

### ISSUE THREAD

**Issue by chriseppstein** [sass/sass#1466]
As a followup to our announcement that Ruby Sass and libsass would have tighter collaboration, I'd like to create a separate repo in the sass organization where we will centralize all new features and changes to Sass: The Language™.

Features there will be closed when the change has been approved and implemented in sass-spec, sass, and libsass.

I'd like to keep the RFC process very informal. Anyone and everyone should be able to follow along and weigh in -- just like our current issue tracker.

As a transition plan, I think we should just link to existing Sass feature requests but any new requests should originate there.

This new repo is where we'd be able to track what features are in what versions of Sass: The Language™ starting with what's in 4.0 (assuming 4.0 is the release after 3.4).

I'd like each RFC to be named for release and detectable using feature detection (aka feature-exists())

cc: @nex3 @hcatlin @akhleung

**Comment by lolmaus** [comment]
Will the new spec accept features requested for years but never implemented?

**Comment by chriseppstein** [comment]
@lolmaus The goal is to converge implementations, not create a loophole for some implementations to add features that aren't in ruby sass.

**Comment by chriseppstein** [comment]
@lolmaus but if those features have been approved, but just not implemented, then yes, once we implement them.

**Comment by HamptonMakes** [comment]
+1

**Comment by HamptonMakes** [comment]
I would recommend that the Sass Reference actually serve as that document. I've noticed a couple issues that aren't actually well defined in the official reference (comment interpolation anyone?) and it might be useful to actually _document_ it first, as part of the proposal, and that way we can make sure that everyone understands before we write specs.

**Comment by chriseppstein** [comment]
Official reference docs will be an outcome of this process. Often ideas need a fair bit of iteration from initial suggestion to finalized feature.

**Comment by HamptonMakes** [comment]
Agreed, which is why I was thinking that using pull requests and iterations
on those might be a nice way to look at a proposal. Often times, feature
issues can just be long series of ideas, disagreements, and can be fairly
fractured. If, after initial discussion, there is a pull request for
someone, and they've written up the idea in a more fleshed out way than
re-reading the issue, it might make the conversations go easier. Imagine a
pull request that defines how a feature works, and after people make
suggestions, it gets changed until we agree and it can be merged into the
new feature branch.

**Comment by nex3** [comment]
I'm not excited about moving feature discussions out of this repository or adding process to them. Doing so adds a lot of work for users, whose instinct will continue to be making feature requests here, and for me, since I'll have to triage all those issues over to the other repository and shepherd them through the process.

Chris, I'm not very worried about keeping features in sync between Ruby Sass, libsass, and sass-spec. I expect to make sass-spec the canonical location of tests once it's on par with the Ruby Sass test suite, and I think doing so will provide plenty of visibility into feature disparity between the implementations. If that ends up not being the case, we can reevaluate, but I'd rather start with a simpler solution.

Hampton, I don't think having a separate space for feature discussions will have the effect you want of making those discussions clearer and more readable. The implementation of a feature influences its design; there's no way to write down what the feature will do before its implemented and expect that to remain consistent. What's more, doing so for all of the [many](https://github.com/sass/sass/labels/Feature) in-discussion features is a lot of work that I don't really have time for.

**Comment by chriseppstein** [comment]
@nex3 Once libsass is at feature parity I think we need a few things to keep things that way:
1. A clear explanation of every language change in english (not just a code delta).
2. Every language change needs to be named so that we can report compatibility on it.
3. A way for libsass to weigh in on whether a language change should be made. If a feature
   is onerous for them to implement or would have undue performance impacts, then we should
   consider that carefully before going forward.

We're keeping two projects in sync now. It's going to involve some process beyond just adding some tests to sass-spec. I don't want the process to be cumbersome, Maybe it's as simple as getting :+1: from the libsass team on the Sass issue tracker to know they've acknoledged a feature is coming and they don't have any major reservations about it

**Comment by nex3** [comment]
I'm not convinced that keeping lockstep parity will need to be a goal after parity is reached. I'm hoping to substantially mitigate the pain of using the canonical implementation with the compilation work I'm doing now, to the extent that libsass will only be necessary for users with truly exceptional performance needs. Allowing the Ruby implementation to stay ahead of libsass will make it substantially easier to quickly iterate on the language.

**Comment by chriseppstein** [comment]
@nex3 So, basically, this nullifies what we said we would do at SassConf. You're saying you're not going to go out of your way to define the language outside of helping write some tests and now you're saying you're not going to wait for them to release. What exactly are we going to do to keep these communities from significantly diverging?

**Comment by chriseppstein** [comment]
@nex3 Another thing to note is that speed is not the only driver of libsass adoption. libsass makes it trivial to use sass in a variety of host languages where exec'ing out to a ruby environment is complex and slow. libsass is the preferred implementation for php, node, java, and perl environments because it's less hassle to install.

**Comment by xzyfer** [comment]
I have to agree with @chriseppstein here. Performance is a reason to use Libsass, but it's far from the only one.

For instance Libsass
- is a better fit for the current generation of node based frontend tooling.
- allows companies who for whatever reason cannot use Ruby in their operating environments. 
- can free developers of what is potentially their only Ruby dependency (and the costs and risk of support another language, and point of failure i.e. rubygems.org)

The final point is the case for my workplace, and many others I'm sure.

**Comment by chriseppstein** [comment]
FWIW, I don't think Ruby Sass and libSass need to ship on the same day. However, I think that ruby sass should never be more than release ahead of libsass. libsass might be a few weeks behind ruby sass.

**Comment by nex3** [comment]
I agreed to a feature freeze until LibSass was entirely compatible with the current release; that still holds. My impression from talking to Hampton is that the biggest cause of divergence is the backlog of compatibility work they had to do on top of handling new features, which will go away once they're caught up.

I have full confidence in their ability to stay in sync once they get there. Despite our best efforts, we don't make releases all that quickly, and even if their lag is on the order of months they'll have plenty of time to get there before we start serious work on the next set of features. But if we hold back or add bureaucracy to the Ruby release in the name of synchronized, then we lose momentum not just on the implementations but on the language itself.

> FWIW, I don't think Ruby Sass and libSass need to ship on the same day. However, I think that ruby sass should never be more than release ahead of libsass.

I propose this, then. Once libsass is compatible with 3.4, we continue on as usual. If we find that the Ruby implementation gets more than one release ahead, I'll reconsider this proposal; if not, you'll let it drop.

Re: using libsass, I'm also keen on getting the setup costs of using Ruby Sass to be far smaller than they are now. Ideally I'd like to distribute a single executable that bundles it entirely for users who don't want to deal with Ruby.

**Comment by chriseppstein** [comment]
> if not, you'll let it drop.

I'm willing to got through a release or two with libsass being in close proximity to the Sass feature set and see how that coordination is going. If there's issues that some process can help us fix, we can and should revisit our language definition process. I want the _least_ amount of process that works but not less than that.
